<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>
<body>
  <script>
    /*
    5.promise 如何串联多个操作任务？
      （1）promise 的 then() 返回一个新的 promise， 可以生成 then() 的链式调用
      （2）通过 then 的链式调用串连多个同步/异步任务

    6.promise 异常传透？
      （1）当使用 promise 的 then 链式调用时，可以在最后指定失败的回调
      （2）前面任何操作出了异常，都会传到最后失败的回调中处理
    
    7.中断 promise 链？
      （1）当使用 promise 的 then 链式调用时，在中间中断，不再调用后面的回调函数
      （2）办法：在回调函数中返回一个 pending 状态的 promise 对象
    */
    new Promise((resolve, reject) => {
      setTimeout(()=> {
        resolve(1)
      })
    }, 1000).then(value => {
      console.log(value)
      // return 2
      // return Promise.resolve(2)
      // throw 2
      // return Promise.reject(2)
    }).then(
      value => {
        console.log(value)
        setTimeout(() => {
          resolve(3)
      }),
      // onReject 不写，相当于下列写法
      // reason => { throw reason }
      reason => Promise.reject(reason)
    }).catch(reason => {
      console.log(reason)
      // throw reason
      // return Promise.reject(reason)

      // 返回一个 pending 的 promise
      // 中断 promise 链
      return new Promise(() => {})
    }).then(
      value => {
        console.log('onResolved3()', value)
      },
      reason => {
        console.log('onRejected2()', reason)
      }
    )
  </script>
</body>
</html>